### PR TITLE
Allows default values in spec to be JSON syntax

### PR DESCRIPTION
**Pull request by dkeeney**
Now an array of values can be configured as the default value for a parameter in the spec.
Allows syntax like  "[123, 5678, 97]" as a default value.

### CODE REVIEW

**Comment by Thanh-Binh** [comment]
@dkeeney thanks for your fixed.
My comment: users may have many mistakes between what is defined in region spec() and what is inputed from JSON. For example, in region spec(), they declare the default value as a scalar, i.e. "1.23" but in JSON they try to input an array.
In this case we should check for data compatibility before parsing and arise an error if it is incompatible.

**Comment by dkeeney** [comment]
Oh, that is a good idea.  I currently check the field names to be sure unexpected parameters are not passed in.  But if a default value is used I should verify that the 'count' property is not a 1 if the result is not a scalar.

**Comment by dkeeney** [comment]
I extended the evaluation of the parameters.  Making sure they were valid parameters, inserting defaults when given, making sure it is an array if the spec indicates an array, and confirming it is not a ReadOnly parameter.  Also extended the unit tests a little.

As it often happens when more testing is performed I found bugs that I did not expect in code not directly related to the problem at hand.  So some of the changes fixed those bugs.

This addresses issue #840 

This broke @Lupino 's API in PR #834 where the api was trying to work around a bug I had with the returned JSON content. This was fixed.  It now returns just the JSON data and not the type, etc.

**Comment by dkeeney** [comment]
Hmmm, got a "ConnectionRefusedError: [Errno 61] Connection refused" while running on MacOS.  Hard to say what is causing this.  Perhaps the server had not yet completely removed its network connection while shutting down before another instance of the server started on the same port.
Perhaps we can alternate port numbers as each test starts its server.

File: py/htm/examples/rest/htm_rest_api.py [hunk 1]
```diff
@@ -1,6 +1,7 @@
 from urllib.parse import urlencode
 import requests
 import json
+import re
```
Comment (Lupino) [inline py/htm/examples/rest/htm_rest_api.py:4]: no use import

File: py/htm/examples/rest/htm_rest_api.py [hunk 2]
```diff
@@ -22,18 +23,12 @@ def request(method, url, data=None, verbose=False):
   if text[0:6] == 'ERROR:':
     raise NetworkRESTError(text[6:].strip())
 
-  if text == 'OK':
+  # determine if result contains JSON i.e. it starts with '{' for a map or '[' for an array
+  if text[0:1] == '[' or text[0:1] == '{':
+    return json.loads( text )
+  else:
     return text
```
Comment (breznak) [inline py/htm/examples/rest/htm_rest_api.py:30]: ok. could even be `try-except` since you return text as fallback anyway (for cases with whitespace, comments etc)

File: src/htm/engine/RegionImpl.cpp [hunk 3]
```diff
@@ -45,21 +44,17 @@ std::string RegionImpl::getName() const { return region_->getName(); }
...
-                << " does not exist in region spec";                              \
-    ParameterSpec p = region_->getSpec()->parameters.getByName(name);          \
-    if (p.dataType != NTA_BasicType_##MethodT)                                 \
-      NTA_THROW << "getParameter" #Type ": Region type " << getType()          \
-	            << ", parameter " << name                                      \
-                << " is of type " << BasicType::getName(p.dataType)            \
-                << " not " #Type;                                              \
-    NTA_THROW << "getParameter" #Type " --  parameter '"       \
-                << name << "' on region of type " << getType() << " not implemented"; \
+#define getParameterInternalT(MethodT, Type)                                                                           \
```
Comment (breznak) [inline src/htm/engine/RegionImpl.cpp:47]: nit, not part of this PR, but could this be a real method, not `define`d?

File: src/htm/regions/TestNode.cpp [hunk 4]
```diff
@@ -35,17 +35,24 @@
 
 namespace htm {
 
-TestNode::TestNode(const ValueMap &params, Region *region)
+TestNode::TestNode(const ValueMap &par, Region *region)
     : RegionImpl(region), computeCallback_(nullptr), nodeCount_(1)
 {
+  outputElementCount_ = 0;
+  spec_.reset(createSpec());
+  ValueMap params = ValidateParameters(par, spec_.get());
```
Comment (breznak) [inline src/htm/regions/TestNode.cpp:43]: is this the new usage? From params, obtain new, validated params by calling `ValidateParameters()`? Couldn't the validation be done implicitly?

File: py/htm/examples/rest/htm_rest_api.py (hunk 2 above)
Comment (dkeeney) [reply py/htm/examples/rest/htm_rest_api.py:30]: > # determine if result contains JSON i.e

I am not happy about this part of the design.  It is still possible that some response that had not intended to be JSON, but pure text, might begin with [ or {.  Leading white space is taken care of because prior to this the value is trimmed. 

Using a try-except would catch most cases where the parser does not recognize a syntax but if there is a real error in the syntax it should throw and not try to use it as text.

 I was thinking that maybe I could rely on the MIME to determine if I should do the JSON decode.  It would depend on the server side knowing what the content contains.

Well, this will do for now but I will think about this.

File: src/htm/engine/RegionImpl.cpp (hunk 3 above)
Comment (dkeeney) [reply src/htm/engine/RegionImpl.cpp:47]: >  but could this be a real method, not defined?

No, I tried every which way and could not make it work as a templated function.  Passing the method as a function pointer we can do but passing the type is hard.

File: src/htm/regions/TestNode.cpp (hunk 4 above)
Comment (dkeeney) [reply src/htm/regions/TestNode.cpp:43]: > Couldn't the validation be done implicitly?

Originally this was performed in the factory before creating the region. but the region implementation had less control.  I struggled with this for quite a while.  I may yet go back to performing it in the factory before creating the region.

File: src/htm/engine/RegionImpl.cpp (hunk 3 above)
Comment (dkeeney) [reply src/htm/engine/RegionImpl.cpp:47]: Actually I have thought about how to re-engineer how arguments are get/set in the region after the region is created.  I just did not want to break the API.

File: py/htm/examples/rest/htm_rest_api.py (hunk 2 above)
Comment (dkeeney) [reply py/htm/examples/rest/htm_rest_api.py:30]: Perhaps we should be consistent and encode ALL response text as JSON, even the OK.  So scalars would be quoted strings.  This is more important now that parameters could be defined as being something more than just a scalar or an array of scalars.  The Value object allows almost any data structure although the getters/setters for parameters do not currently support it.  If we make this all JSON it will allow future changes to the parameter handling.

**Comment by dkeeney** [comment]
In conclusion, I will do the following:
1) change all REST responses to be strict JSON but this should be a separate PR.
2) leave discussions about the parameters for later because this could cause a major impact on the API.

File: src/htm/engine/RegionImpl.cpp (hunk 3 above)
Comment (breznak) [reply src/htm/engine/RegionImpl.cpp:47]: > Actually I have thought about how to re-engineer how arguments are get/set in the region after the region is created. I just did not want to break the API. 

maybe we can start thinking about NAPIv2 (along with the cleanup of obsoleted method + providing the RESTful interface) .
